### ISSUE TITLE
Format: changes to `out_functions` during pretty-printing are not used.

### ISSUE THREAD

**Issue by dbuenzli** [ocaml/ocaml#13371]
This may be turned into a bug or a documentation improvement.

I want to update a formatter's newline output while formatting is being done in a particular box[^1]. I thought the following combinator that takes a pretty printer and changes the newline output function during the call to the pretty printer would do:

```ocaml
let with_newline ~nl pp = fun ppf v ->
  let outf = Format.pp_get_formatter_out_functions ppf () in
  let out_newline () = outf.out_string nl 0 (String.length nl) in
  let outf_with_nl = { outf with out_newline } in
  let finally () = Format.pp_set_formatter_out_functions ppf outf in
  Format.pp_set_formatter_out_functions ppf outf_with_nl;
  Fun.protect ~finally @@ fun () -> pp ppf v
```

But that doesn't work: 

```ocaml
> ocaml
OCaml version 5.2.0
# let pp_lines ppf ls = Format.fprintf ppf "@[<v>%a@]" (Format.pp_print_list Format.pp_print_string) ls;;
# let pp_breaks = with_newline ~nl:" \\\n" pp_lines;;
# Format.eprintf "%a@." pp_breaks ["hey"; "ho"];;
hey
ho
```

I suspect the `out_functions` get cached somewhere before pretty-printing starts. Either: 

1. This can be fixed.
2. Works as intended. In which case it could be perhaps added in the doc of 
   `pp_set_formatter_out_functions` that calling this function during 
   pretty-printing has no effect.

[^1]: Use case: in a box that prints a shell command, lines should be terminated by `\` before the newline character so that the command can be easily cut and pasted.

**Comment by jberdine** [comment]
The side-effects of `Format.pp_set_formatter_out_functions` are not synchronized with the time when the printing functions are actually called as strongly as your code would require. The queue internal to `Format` can still contain pending items to print when the `pp` argument to `with_newline` returns, so that when the queue is later flushed by `@.`, the `outf` function has already been reset. If you add `@.` to the format string in `pp_lines`, then the overridden function is called. In this sort of situation, I have used the `kprintf` functions to schedule the resetting of the printing functions to make sure they happen at the desired time.

**Comment by dbuenzli** [comment]
Thanks for your answer @jberdine. I guess you suggest something like: 
```ocaml
  let with_newline ~nl pp = fun ppf v ->
    let outf = Format.pp_get_formatter_out_functions ppf () in
    let out_newline () = outf.out_string nl 0 (String.length nl) in
    let outf_with_nl = { outf with out_newline } in
    let finally ppf =
      Format.pp_print_flush ppf ();
      Format.pp_set_formatter_out_functions ppf outf
    in
    Format.pp_set_formatter_out_functions ppf outf_with_nl;
    Format.kfprintf finally ppf "%a" pp v
```

The problem of course is that flushing destroys all the boxing context. Somehow you would like to change the end of lines only for those newlines that are going to happen because of `pp`. I guess if `Format`'s render queue has no "state change" elements, it's hopeless.

For now I solved my problem by formatting to a string and then formatting the lines but it's not very compositional, e.g. doesn't take into account the current indentation level.

**Comment by jberdine** [comment]
Yes, the state of the out functions and the formatting queue are not correlated internally, so the only reliable points to make such changes is immediately following a flush. One not very efficient and not very satisfying workaround I have used for something similar is [here](https://github.com/microsoft/SLAyer/blob/master/src/Log.ml#L26-L57). That code uses an external form of indentation and wraps `printf` to check if indentation has returned to 0 and if so, flush the formatter to a buffer, then reformat the contents. It is not very nice. Since that code was written, `Format` gained `out_indent`, so it may be possible to set that to a function to track the current indentation level, and test that to determine whether to flush to a buffer.

But overall, the interface allowing to set the output functions seems to be intended for initial configuration prior to starting to format, rather than transient changes during the formatting process.

**Comment by dbuenzli** [comment]
Perhaps it's not the best way to go around this anyways. But I think  `Format` is missing compositional `Format.{prefix,suffix}_lines` combinators. It's not the first time I wished for them.

In any case I think this issue should then be about documenting that as you say: 

> But overall, the interface allowing to set the output functions seems to be intended for initial configuration prior to starting to format, rather than transient changes during the formatting process.

**Comment by nojb** [comment]
Discussed during today's triage meeting and we agreed that this issue should be documented.
